File: kotlin/src/Factorization.kt
fun factorize(N: Long): Map<Long, Int> {
    val factors = mutableListOf<Long>()
    var n = N
    var d: Long = 2
    while (d * d <= n) {
        while (n % d == 0L) {
            factors += d
            n /= d
        }
        ++d
    }
    if (n > 1) factors += n
    return factors.groupBy { it }.mapValues { it.value.size }
}

// Usage example
fun main(args: Array<String>) {
    println(factorize(36))
    println(factorize(150))
}

File: cpp/numbertheory/factorization.cpp
#include <bits/stdc++.h>

using namespace std;

// returns prime_divisor -> power
// O(sqrt(n)) complexity
map<long long, int> factorize(long long n) {
    map<long long, int> factors;
    for (int d = 2; (long long)d * d <= n; d++) {
        while (n % d == 0) {
            ++factors[d];
            n /= d;
        }
    }
    if (n > 1) {
        ++factors[n];
    }
    return factors;
}

// usage example
int main() {
    map<long long int, int> factors = factorize(4 * 3 * 1000000000039);
    for (auto e : factors) {
        cout << e.first << "^" << e.second << " " << endl;
    }
}
